### PR TITLE
[SPARK-45810][Python] Create Python UDTF API to stop consuming rows from the input table

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR creates a Python UDTF API to stop consuming rows from the input table.

If the UDTF raises a `SkipRestOfInputTableException` exception in the `eval` method, then the UDTF stops consuming rows from the input table for that input partition, and finally calls the `terminate` method (if any) to represent a successful UDTF call.

For example:

```
@udtf(returnType="total: int")
class TestUDTF:
    def __init__(self):
        self._total = 0

    def eval(self, _: Row):
        self._total += 1
        if self._total >= 3:
            raise SkipRestOfInputTableException("Stop at self._total >= 3")

    def terminate(self):
        yield self._total,
```

### Why are the changes needed?

This is useful when the UDTF logic knows that we don't have to scan the input table anymore, and skip the rest of the I/O for that case.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @ueshin @allisonwang-db

File: python/pyspark/worker.py [hunk 1]
```diff
@@ -995,6 +995,8 @@ def verify_result(result):
             def func(*a: Any) -> Any:
                 try:
                     return f(*a)
+                except StopIteration:
+                    raise
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply python/pyspark/worker.py:999]: I looked at this, but unlike scalar pyspark UDFs, we want to apply custom treatment for UDTFs with respect to the `eval` vs. `terminate` vs. `cleanup` methods. The idea is that if `eval` invokes `raise StopIteration()`, we then call `terminate` as normal and return a successful result for the UDTF as a whole.

To implement this behavior, if we used `fail_on_stopiteration` here, we'd have to later catch a `PySparkRuntimeException` and check if it's `error_class="STOP_ITERATION_OCCURRED"` manually, which seems pretty confusing. It seems simpler to just `raise` here and catch the `StopIteration` on L1062 and L1154 instead, and then call `terminate`.

**Review by dtenedor — COMMENTED** [review]
Thanks @allisonwang-db for your review!

File: python/pyspark/worker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (allisonwang-db) [reply python/pyspark/worker.py:999]: Make sense!

File: python/pyspark/worker.py [hunk 2]
```diff
@@ -1057,6 +1059,9 @@ def mapper(_, it):
                     yield from eval(*[a[o] for o in args_kwargs_offsets])
                 if terminate is not None:
                     yield from terminate()
+            except StopIteration:
+                if terminate is not None:
+                    yield from terminate()
```
Comment (allisonwang-db) [inline python/pyspark/worker.py:1064]: If we are catching the `StopIteration` exception in mapper, do we still need the try ... catch block inside the `def func` and `def evaluate` below?

Comment (allisonwang-db) [reply python/pyspark/worker.py:1064]: Ah the exception thrown in eval will be caught by this 
```
except Exception as e:
    raise PySparkRuntimeError(
       error_class="UDTF_EXEC_ERROR",
```
I wonder if we should just move the whole try ... except ... block to this mapper function instead of checking this for every single input row.

Comment (dtenedor) [reply python/pyspark/worker.py:1064]: I tried this, but then we lose the UDTF method name (i.e. `eval` or `terminate`) in the error message. It seems better to keep that. I will leave this alone for now, let me know if you would prefer to continue with this change anyway and lose this information in the error message and we can proceed with that.

**Review by dtenedor — COMMENTED** [review]
Thanks @allisonwang-db  for your review!

**Comment by dtenedor** [comment]
cc @ueshin

**Comment by ueshin** [comment]
btw, the example in the description should not use tabs? It breaks the code format.

File: python/pyspark/sql/tests/test_udtf.py [hunk 3]
```diff
@@ -2467,6 +2468,41 @@ def terminate(self):
...
+        @udtf
+        class TestUDTF:
+            def __init__(self):
+                self._total = 0
+
+            @staticmethod
+            def analyze(_):
+                return AnalyzeResult(
+                    schema=StructType().add("total", IntegerType()), withSinglePartition=True
+                )
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py]: Good idea, done!

File: python/pyspark/sql/tests/test_udtf.py [hunk 4]
```diff
@@ -2467,6 +2468,41 @@ def terminate(self):
...
+                """
+                WITH t AS (
+                  SELECT id FROM range(1, 21)
+                )
+                SELECT total
+                FROM test_udtf(TABLE(t))
+                """
+            ),
+            [Row(total=4)],
+        )
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py:2500]: The `SkipRestOfInputTableException` stops scanning rows for just the current partition. I added a test case for this as well.

File: python/pyspark/sql/udtf.py [hunk 5]
```diff
@@ -118,6 +125,13 @@ class AnalyzeResult:
     orderBy: Sequence[OrderingColumn] = field(default_factory=tuple)
 
 
+# This represents an exception that the 'eval' method may raise to indicate that it is done
+# consuming rows from the current partition of the input table. Then the UDTF's 'terminate' method
+# runs (if any).
+class SkipRestOfInputTableException(Exception):
+    pass
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply python/pyspark/sql/udtf.py]: Good point, I moved this to a pydoc string inside the class definition itself.

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin for your review! Addressed your comments, please take another look.

File: python/docs/source/user_guide/sql/python_udtf.rst [hunk 6]
```diff
@@ -101,20 +101,29 @@ To implement a Python UDTF, you first need to define a class implementing the me
...
             Examples
             --------
-            analyze implementation that returns one output column for each word in the input string
-            argument.
+            This is an `analyze` implementation that returns one output column for each word in the
+            input string argument.
 
-            >>> def analyze(self, text: str) -> AnalyzeResult:
+            >>> @staticmethod
+            ... def analyze(text: str) -> AnalyzeResult:
```
Comment (ueshin) [inline python/docs/source/user_guide/sql/python_udtf.rst:117]: Thanks for the fix! 👍

File: python/pyspark/sql/tests/test_udtf.py [hunk 7]
```diff
@@ -2467,6 +2468,53 @@ def terminate(self):
...
+                """
+                WITH t AS (
+                  SELECT id FROM range(1, 21)
+                )
+                SELECT id / 10 AS id_divided_by_ten, total
+                FROM test_udtf(TABLE(t) PARTITION BY id / 10)
+                ORDER BY ALL
+                """
+            ),
+            [Row(id_divided_by_ten=0, total=4), Row(id_divided_by_ten=1, total=4)],
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_udtf.py:2515]: For this case, the expected results should be:

```py
[Row(id_divided_by_ten=0, total=4), Row(id_divided_by_ten=1, total=4), Row(id_divided_by_ten=2, total=1)]
```

Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py:2515]: I thought so as well, but apparently the `range` function accepts the second argument for its upper bound (exclusive) :) 

```
> SELECT id FROM range(1, 21)

id
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
```

Comment (ueshin) [reply python/pyspark/sql/tests/test_udtf.py:2515]: so `20` should be in another group?

If the stop is not here, there will be 3 groups including 9, 10, and 1 rows.

File: python/pyspark/sql/tests/test_udtf.py [hunk 8]
```diff
@@ -2467,6 +2468,53 @@ def terminate(self):
...
+        # SkipRestOfInputTableException stops scanning rows for each of the two partitions
+        # separately.
+        assertDataFrameEqual(
+            self.spark.sql(
+                """
+                WITH t AS (
+                  SELECT id FROM range(1, 21)
+                )
+                SELECT id / 10 AS id_divided_by_ten, total
+                FROM test_udtf(TABLE(t) PARTITION BY id / 10)
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_udtf.py:2511]: btw, `id / 10` will be `double`. I guess it should be `floor(id / 10)`?

Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py]: Sounds good, done.

File: python/pyspark/sql/tests/test_udtf.py (hunk 7 above)
Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py]: You're right, this was actually a bug. The `class UDTFWithPartitions` did not have support for this new exception type before. I added support for that now, and the test passes.

File: python/pyspark/sql/tests/test_udtf.py [hunk 9]
```diff
@@ -2482,6 +2533,7 @@ def tearDownClass(cls):
             super(UDTFTests, cls).tearDownClass()
 
 
+'''
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_udtf.py:2536]: Looks like a mistake? Could you revert this change?

File: python/pyspark/sql/tests/test_udtf.py [hunk 10]
```diff
@@ -2775,7 +2827,7 @@ def tearDownClass(cls):
             cls.spark.conf.unset("spark.sql.execution.pythonUDTF.arrow.enabled")
         finally:
             super(UDTFArrowTests, cls).tearDownClass()
-
+'''
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_udtf.py:2830]: ditto.

File: python/pyspark/sql/tests/test_udtf.py (hunk 9 above)
Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py]: My mistake on this, reverted.

File: python/pyspark/sql/tests/test_udtf.py (hunk 10 above)
Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py]: My mistake on this, reverted.

**Review by ueshin — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.
